pkg/cloudprovider/aws/fleet/instance.go

```diff
@@ -78,3 +81,30 @@ func (p *InstanceProvider) Create(ctx context.Context,
 	}
 	return createFleetOutput.Instances[0].InstanceIds[0], nil
 }
+
+func (p *InstanceProvider) getInstanceIDs(nodes []*v1.Node) []*string {
+	ids := []*string{}
+	for _, node := range nodes {
+		id := strings.Split(node.Spec.ProviderID, "/")
+		if len(id) < 5 {
+			zap.S().Debugf("Continuing after failure to parse instance id, %s has invalid format", node.Name)
+			continue
+		}
+		ids = append(ids, aws.String(id[4]))
+	}
+	return ids
+}
+
+func (p *InstanceProvider) TerminateInstances(ctx context.Context, nodes []*v1.Node) (*int, error) {
+	ids := p.getInstanceIDs(nodes)
+
+	output, err := p.ec2.TerminateInstancesWithContext(ctx, &ec2.TerminateInstancesInput{
+		DryRun:      aws.Bool(false),
+		InstanceIds: ids,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("terminating %d instances, %w", len(output.TerminatingInstances), err)
+	}
+
+	return aws.Int(len(output.TerminatingInstances)), nil
```

Comment:
I'd avoid returning an int here.

Reply:
As opposed to what, nothing?

Reply:
I thought I left a comment on this elsewhere, but I'll cover it here. Returning the number of successful nodes is either too much or too little information to pass through this API. Returning only the number is useful in exactly 1 use case (yours): logging the number of failures that happened. 

If you wanted to implement sophisticated retry logic, you'd need to pass at least instance ID so that Karpenter can actually do something. More effective would probably be the list of nodes that failed.

We need to think carefully about what this failure mode is (EC2 failing to delete *some* of the nodes), how we can tolerate it, and how we can avoid leaking EC2 specific failures through the cloud provider API.

Keep in mind, the cloudprovider is one of the most important APIs in the Karpenter project. It represents the clean line between the AWS and Kubernetes worlds. We want to be really careful that the interface makes sense at a theoretical and conceptual level, such that one cloud provider's quirks doesn't pollute others. In this case, I'd consider EC2's failure mode of only deleting some instances as a cloud provider quirk -- it's a really awkward failure mode. 

I actually think it's fine if our cloud provider simply drops a debug log that it's failed, and Karpenter can keep trying to delete the instance on the next loop. This is actually the exact same behavior as your current implementation, except it doesn't pollute the cloud provider API.

The other way to approach this is to try to build a robust backoff-retry strategy within Karpenter's reallocator, which as mentioned above, would require passing more sophisticated information than just the failure count.

Reply:
You bring up some good points. Thanks for taking your time to type this out.